llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -0,0 +1,296 @@
+//===----- SpecConstants.cpp - SYCL Specialization Constants Pass ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "SpecConstants.h"
+
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+
+using namespace llvm;
+
+namespace {
+
+// __sycl* intrinsic names are Itanium ABI-mangled; this is common prefix for
+// all mangled names of __sycl_getSpecConstantValue intrinsics, which differ by
+// the template type parameter and the specialization constant value type.
+constexpr char SYCL_GET_SPEC_CONST_VAL[] = "_Z27__sycl_getSpecConstantValue";
+// Unmangled base name of all __spirv_SpecConstant intrinsics which differ by
+// the value type.
+constexpr char SPIRV_GET_SPEC_CONST_VAL[] = "__spirv_SpecConstant";
+// Metadata ID string added to calls to __spirv_SpecConstant to record the
+// original symbolic spec constant ID.
+constexpr char SPEC_CONST_SYM_ID_MD_STRING[] = "SYCL_SPEC_CONST_SYM_ID";
+
+static void AssertRelease(bool Cond, const char *Msg) {
+  if (!Cond)
+    llvm_unreachable((Twine("SpecConstants.cpp: ") + Msg).str().c_str());
+}
```

Comment:
I did not get what is the purpose of this function. `llvm_unreachable` and standard c asserts work in the same way, i.e. in debug mode of when it's "build with asserts" configuration. In other words, using `llvm_unreachable" won't make it check something in release mode unless asserts are enabled.
My point was that we are doing some assertions only under
`#ifndef NDEBUG` so they are not presented in the code for `release with asserts` configuration.

Reply:
Right, remnant from experiment. Will change to using llvm_unreachable directly.

Reply:
> Will change to using llvm_unreachable directly.

I need to take this back, see below.

> In other words, using llvm_unreachable won't make it check something in release mode unless asserts are enabled.

Actually, the experiment showed that -DCMAKE_BUILD_TYPE=Release does set NDEBUG, and the effect of LLVM unreachable on Linux is a signal caught by handler - i.e. a crash. So it does work as release assert. And I don't want to write `if (Cond) llvm_unreachable();` everywhere in the code.

Reply:
BTW, this indeed diverges from what https://llvm.org/docs/CodingStandards.html says: 
> Neither assertions or llvm_unreachable will abort the program on a release build.'

Reply:
> Actually, the experiment showed that -DCMAKE_BUILD_TYPE=Release does set NDEBUG, and the effect of LLVM unreachable on Linux is a signal caught by handler - i.e. a crash. So it does work as release assert.
 
This is really strange! Assertions & llvm_unreachable should be triggered in Release build only if `LLVM_ENABLE_ASSERTIONS` cmake variable is set to `ON`...

Reply:
Even the doc for `llvm_unreachable` says that it gives errors only in debug mode, https://llvm.org/doxygen/Support_2ErrorHandling_8h.html#ace243f5c25697a1107cce46626b3dc94

Reply:
So:
- description in the docs does not match the  behavior
- behavior is "release assert"
- what actually was under question is just the name of the function, as LLVM anyway recommends unreachable over assert, and if it did not give crash in release, it would have been called CheckAndAssert or somthing.

Given that I believe the code should be left as is here

Reply:
Please see below - replaced with report_fatal_error